### ISSUE TITLE
[Call for Bench] SWE-Paddle 社区共建

### ISSUE THREAD

**Issue by sunzhongkai588** [PaddlePaddle/Paddle#79447]
> [!IMPORTANT]
> 简单来说：把经合入 Paddle 的 PR，整理成可复现、可验证的真实研发任务，作为高质量 benchmark 用来评测大模型，拿到奖励与论文署名机会。
> 
> SWE-Paddle 采用 **proposal-first** 的共建流程：贡献者先在 `PaddlePaddle/community` 仓库下提交一个轻量 `proposal.md`；proposal 被 approve 后，再在同一目录继续补齐完整任务包。这样可以避免大家一开始就拆 patch、补环境、写完整说明，最后却发现样本不适合收录。

---

### SWE 类 benchmark 是什么

早期的代码评测（HumanEval、MBPP）让模型在“写孤立函数题”，这类评测往往有以下问题：题目短、没有上下文、和真实开发差得很远。

2023 年的 **SWE-bench** 换了个思路：丢给模型一个**真实开源仓库**和一个**真实 issue**，让它像工程师一样在整个代码库里定位问题、跨文件改代码、最后让**测试通过**。它考的不再是"能不能写对一个函数"，而是"**能不能解决一个真实的工程问题**"——这也是目前衡量 coding agent 最有说服力的方式。

但 **SWE-bench** 也存在很多问题，基于此发布了很多衍生benchmark，如：

- **SWE-bench Verified**：发现 **SWE-bench**  存在诸多现象——题目泄露答案、测试无效、描述不清。于是 OpenAI 从原集筛出 500 条**人工确认**的高质量样本。

- **Multi-SWE-bench**：发现 **SWE-bench** 只测 Python 覆盖面太窄，扩到多语言、多仓库。

- **SWE-bench Pro**：发现 **SWE-bench Verified** 公开榜单已经被刷烂、模型已经学会"背题"，于是改用商业化的仓库，做更难、更长程、抗污染的任务。

因此，对于 SWE 系列的改进，行业内一直致力于构建高质量的数据集——这取决于数据集是否足够贴近真实研发流程、任务难度是否足够有挑战性、测试是否能够很好的衡量大模型生成的代码。但不管怎样，一条 SWE 任务样本所需要的内容都包含以下：

- **issue（问题描述）** ：给 agent 看的题目（可以是 bug、缺失的 API、要对齐的行为等）。

- **test patch（测试补丁）**：描述目标行为的测试——代码改动前测试必须 **fail**（bug 未修 / API 未加时它就该失败）。

- **gold patch（标准答案）**：来自已合入 PR 的真实改动——打上后测试必须 **pass**。

此外，加上代码改动前的 `base_commit` + 可复现环境，以便于复现

---
用 SWE-Paddle 里的真实样本 **Paddle #70102**（`paddle.nn.functional.fold` 边界 bug）串起来看：

- **issue**：`fold` 在 `output_sizes` 含 1 时（如 `output_height=1`）会错误报错 `output_height should be greater than one`，和 PyTorch 行为不一致。要求修掉这个尺寸校验，但**不放宽**其他非法输入的校验。
- **test patch**：新增 `output_height=1` 的边界用例，打到 base_commit 上 → 报错 **fail**（bug 被暴露）。
- **gold patch**：来自 PR #70102 的真实修复（只改 infermeta 里的尺寸校验分支）→ 打上后目标测试 **pass**。

三者配上 `base_commit`（`49a6b0a4...`）和可复现环境，就构成一条能自动判分的样本。**这就是各位要产出的东西。**

```text
问题描述 + base_commit + 测试补丁 + gold patch + 可复现环境
```

---

### 为什么做 Paddle 的 SWE benchmark？

- **主流榜单几乎没有 Paddle**：SWE 系列覆盖了一大批 Python / Java / Go 应用型项目，但**深度学习框架、尤其是 Paddle，几乎是空白**。这块空白既是缺口，也是我们的机会。

- **Paddle PR 天然是高质量数据**：Paddle 的合入 PR 有一套成熟流程的约束：

  - PR 单测覆盖率要求高（**> 90%**）；

  - 每个 PR 至少经过 **2 位研发 review**；

  - 有回归约束和 CI gate。

换句话说，Paddle 社区已经用真实工程流程，做完了最宝贵的那部分"标注"——真实需求、真实代码上下文、真实测试约束、真实  reviewer 的隐含标准。我们要做的，只是把它转成标准、可复现、可评测的benchmark。

- Paddle 任务有较大的区分度：我们尝试根据以往的黑客松，构造了 10 条数据，并测量了 3 个不同模型的效果，发现了 SWE-Paddle 确实能**拉开模型差异**

| 模型            | 通过   | 成功率 |
| --------------- | ------ | ------ |
| Claude Opus 4.8 | 7 / 10 | 70%    |
| DeepSeek-V4 Pro | 4 / 10 | 40%    |
| ERNIE-5.1       | 3 / 10 | 30%    |

 

---

## 2. 谁可以参与

> [!IMPORTANT]
> 只要你是 Paddle contributor（给 Paddle 提过并合入过 PR），就可以参与。

我们欢迎两类角色（可以兼任），做的事不同，但都是正式贡献、都计入贡献榜与论文贡献者：

- benchmark 数据贡献者（主要）
- reviewer / 环境贡献者

### （1）数据贡献者

只要你给 Paddle 提过并合入过 PR，就可以把自己熟悉的 PR 转成 SWE-Paddle 任务候选。
  
优先推荐：
  - 你自己提交并合入的 PR；
  - 你 review 过、很熟悉背景的 PR；
  - 你熟悉模块里的历史 PR；
  - 已修复且能对应到明确 PR 的真实 issue。
  
不建议提交：  
  - 未合入 PR；
  - 纯文档、格式化、重命名类 PR；
  - 没有测试、无法自动判分的 PR；
  - 多个独立目标混在一起、无法拆分的 PR；
  - 依赖不可固定外部服务或临时下载链接的 PR；
  - 题面会明显泄露答案路径的 PR。

### （2）reviewer / 环境贡献者

如果你很熟悉 Paddle 开发，也可以参与 review 和环境验证：
  - **帮忙 review 样本**：判断 patch 是否合理、题面有没有泄露答案、测试是否有效、难度是否名副其实（熟悉 Paddle 相关模块的同学尤其合适）
  - **帮忙构建 / 验证环境**：构建环境、复现环境非常麻烦，需要有相关方面知识的开发

---

## 3. 提交到哪里

统一提交到 [PaddlePaddle/community](https://github.com/PaddlePaddle/community/tree/master/swe-paddle) 仓库 `swe-paddle` 目录下：

```text
swe-paddle/
  README.md
  proposal_template.md
  tasks/
    PaddlePaddle__Paddle-68432/
      proposal.md
```

原始 Paddle PR 只作为来源链接引用，不需要在 `PaddlePaddle/Paddle` 主仓重复开 issue。

当前已经准备了一个 starter sample：

```text
swe-paddle/tasks/PaddlePaddle__Paddle-68432/
```

这个目录展示了 proposal approve 后的完整形态。

---

### Step 1：先提交 proposal

贡献者先在 `swe-paddle/tasks/` 下新建目录，只提交 `proposal.md`：

```text
swe-paddle/
  tasks/
    PaddlePaddle__Paddle-xxxxx/
      proposal.md
```

目录命名建议：

```text
PaddlePaddle__Paddle-68432
```

其中 `68432` 是原始 Paddle PR 编号。

`proposal.md` 建议使用中文填写，方便 Paddle contributor 和 reviewer 快速理解。必要字段包括：

- PR 链接；
- PR 标题；
- `base_commit`；
- merged 时间；
- 你的身份；
- 问题一句话；
- 为什么适合作为 SWE-Paddle 样本；
- 任务类型和模块标签；
- 验证思路；
- 环境与资源；
- 风险自查。

模板见：

```text
swe-paddle/proposal_template.md
```

### Step 2：维护组 review proposal

维护组会先判断：

- 这个 PR 是否真实、已合入、可追溯；
- 任务目标是否单一清楚；
- 测试是否有希望构造出 fail-before / pass-after；
- 是否存在明显答案泄露风险；
- 环境成本是否可接受；
- 是否适合作为第一批公开样本。

结论可能是：

```text
approved / needs-info / not-fit / deferred
```

### Step 3：proposal approved 后补完整任务包

proposal 通过后，再在同一目录补齐完整任务包：

```text
swe-paddle/
  tasks/
    PaddlePaddle__Paddle-68432/
      README.md
      proposal.md
      instruction.md
      environment/README.md
      solution/code.patch
      tests/test.patch
      tests/test.sh
```

首版先保持目录简洁，不强制提交 `swe_paddle.yaml` 或 `author_checklist.md`。后续如果 verifier、看板或数据发布流程需要机器可读字段，再补轻量 metadata。

---

## 5. 完整任务包里每个文件做什么

- `README.md`：说明这个样本的来源、价值、目录结构和验证入口。
- `proposal.md`：已 approve 的提案，保留上下文，方便 reviewer 回看。
- `instruction.md`：给 coding agent 的自包含题面，不能泄露答案。
- `environment/README.md`：说明 base commit、环境、依赖、运行顺序和已知风险。
- `solution/code.patch`：来自已合入 PR 的真实修复 patch，也就是 gold patch。
- `tests/test.patch`：暴露目标行为的测试 patch。
- `tests/test.sh`：最小目标测试命令。

其中最核心的是三件套：

- **instruction**：模型看到的问题描述；
- **test patch**：修复前应失败、修复后应通过的目标测试；
- **gold patch**：真实修复。

配合 `base_commit` 和可复现环境，就能构成一条可自动判分的 SWE-Paddle 样本。

---

## 6. 环境怎么处理

Paddle 环境比较重，所以首版不强制要求贡献者自己提交 Docker 或镜像（能提供更好）。贡献者需要尽量写清楚环境线索：

- PR 当时或你验证过的 Paddle 来源；
- Python / OS / CUDA / cuDNN 等版本；
- CPU / GPU / 多卡等硬件需求；
- patch 是否涉及 Python、C++、CUDA；
- 最小测试命令；
- 已知 flaky、seed、外部数据依赖；
- 如果有本地 oracle 日志，可以附关键片段。

不确定最低硬件要求时，写“已验证环境 + 未确认最低要求”，不要只写“需要 GPU”。

维护组后续负责构建 era-matched 环境，并运行 Run/Test/Fix verifier。

---

## 7. 怎么验收

核心验收逻辑是 Run/Test/Fix：

```text
Run:  base_commit                              检查基线环境和存量测试
Test: base_commit + test_patch                 增目标测试应 fail / error，已有稳定测试不应新增异常
Fix:  base_commit + test_patch + code_patch     新增目标测试应 pass，已有稳定测试仍应 pass
```

同时还要检查：

- 题面是否自包含；
- 是否泄露答案；
- 测试是否真能暴露问题；
- gold patch 是否真的修复；
- 是否存在 flaky；
- 是否破坏同文件或同模块的存量测试。

正式进入 verified split 前，需要维护组 verifier 和双 reviewer 确认。

---

## 8. 奖励与回报

  - 每条 accepted + verified instance：50 元小礼品。
  - 高质量贡献者：进入 SWE-Paddle 论文署名（有机会投顶会）

如有意向，扫码进如流群交流～
<img width="1170" height="1339" alt="Image" src="https://github.com/user-attachments/assets/a1198aba-f5d0-4234-a0fc-31394caa6387" />
